MDEV-18827 Create utility to parse frm files and print their DDL

hp77-creator opened the pull request:
<!--
Thank you for contributing to the MariaDB Server repository!

You can help us review your changes faster by filling in this template <3

If you have any questions related to MariaDB or you just want to hang out and meet other community members, please join us on https://mariadb.zulipchat.com/ .
-->

<!--
If you've already identified a https://jira.mariadb.org/ issue that seems to track this bug/feature, please add its number below.
-->

## Add FRM parser as a standalone utility to mariaDB server
- [x] The Jira issue number for this PR is: [MDEV-18827](https://jira.mariadb.org/browse/MDEV-18827)

<!--
An amazing description should answer some questions like:
1. What problem is the patch trying to solve?
2. If some output changed that is not visible in a test case, what was it looking like before the change and how it's looking with this patch applied?
3. Do you think this patch might introduce side-effects in other parts of the server?
-->
## Description
The following tests are going to be added and run with mtr:
- [x]  simple columns, at least int, text, varchar(value)
- [x]  keys, `using btree`
- [ ]  keys `using rtree`
- [x]  sinple vcol expressions: `a int, x int as (a *2)` 
- [x]  same for DEFAULT: `a int default (whatever)`
- [x]  DEFAULT CURRENT_TIMESTAMP is a special case
- [x]  field-level comments, table-level comments

==================
- [x]  innodb engine
- [X]  more exotic engines
- [X]  engine options
- [ ] partitions -- currently failing so added with error clause

==================
- [ ] Expressions from plugins, like json, Item_geometry_func, inet6 -- currently failing so added with error clause
- [x] Sequences: default(nextval(s))

==================
- [x] Foreign keys in innodb -- foreign key constraints are not printed properly
## Release Notes
TODO: What should the release notes say about this change?
Include any changed system variables, status variables or behaviour. Optionally list any https://mariadb.com/kb/ pages that need changing.

## How can this PR be tested?

Consult the documentation on ["Writing good test cases"](https://mariadb.org/get-involved/getting-started-for-developers/writing-good-test-cases-mariadb-server).
<!--
In many cases, this will be as simple as modifying one `.test` and one `.result` file in the `mysql-test/` subdirectory.
Without automated tests, future regressions in the expected behavior can't be automatically detected and verified.
-->

run 
```
> ./mtr --suite=client --do-test=mariadb-frm
```

If the changes are not amenable to automated testing, please explain why not and carefully describe how to test manually.

<!--
Tick one of the following boxes [x] to help us understand if the base branch for the PR is correct.
-->
## Basing the PR against the correct MariaDB version
- [x] *This is a new feature or a refactoring, and the PR is based against the `main` branch.*
- [ ] *This is a bug fix, and the PR is based against the earliest maintained branch in which the bug can be reproduced.*

<!--
  All code merged into the MariaDB codebase must meet a quality standard and coding style.
  Maintainers are happy to point out inconsistencies but in order to speed up the review and merge process we ask you to check the CODING standards.
-->
## PR quality check
- [x] I checked the [CODING_STANDARDS.md](https://github.com/MariaDB/server/blob/-/CODING_STANDARDS.md) file and my PR conforms to this where appropriate.
- [x] For any trivial modifications to the PR, I am ok with the reviewer making the changes themselves.

hp77-creator commented on extra/frm_parser.cc:
```diff
@@ -0,0 +1,373 @@
+  {
+    my_free((void*)db_name.str);
+  }
+  if (table_name.str)
```
understood, will add this change

hp77-creator replied:
done, so I will only keep check for `share`

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/MariaDB/server?pullRequest=4094) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/MariaDB/server?pullRequest=4094) before we can accept your contribution.<br/>**3** out of **4** committers have signed the CLA.<br/><br/>:white_check_mark: ParadoxV5<br/>:white_check_mark: hp77-creator<br/>:white_check_mark: apostolis1<br/>:x: Thirunarayanan<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/MariaDB/server?pullRequest=4094) it.</sub>

FooBarrior commented on extra/mariadbfrm/mariadb_frm.cc:
```diff
@@ -20,18 +20,79 @@
+class LEX;
+class Transaction_state;
+
+class THD {
```
No, You can't add THD like this. It'll be available only for mariadb_frm.cc. Other modules will not see it like this. Think of what will happen with access to some field if it has a different layout in  another module's function!

FooBarrior commented on extra/mariadbfrm/mariadb_frm.cc:
```diff
@@ -0,0 +1,531 @@
+static handlerton mock_hton;
+
+static st_maria_plugin mock_plugin = {
+  .type = MYSQL_STORAGE_ENGINE_PLUGIN,
```
Formally, designated initialization is only possible since c++20. I think nothing can stop us to enable it in cmake. Please try to make it, but only for frm_parser!

FooBarrior commented on extra/mariadbfrm/mariadb_frm.cc:
```diff
@@ -0,0 +1,531 @@
+  
+  frm_data= read_frm_file(frm_path, &frm_length);
+  if (!frm_data)
+    goto cleanup;
```
We have different opinions about gotos acros our communities, and mine is -- they are completely unnecessary, they make the code harder to read, and they make it impossible to add a variable initialization in-between the function (after a first goto).

Try to rewrite cleanups with the use of RAII. We have an ATEXIT convenience macro for that.
THD uses constructor and destructor, so use  unique_ptr for a THD pointer.

FooBarrior commented on extra/mariadbfrm/mariadb_frm.cc:
```diff
@@ -0,0 +1,531 @@
+  printf("DEBUG: Allocating TABLE_SHARE manually\n");
+  fflush(stdout);
+  
+  share = (TABLE_SHARE*)my_malloc(PSI_NOT_INSTRUMENTED, sizeof(TABLE_SHARE), MYF(MY_ZEROFILL));
```
No need. TABLE_SHARE is small enough to allocate it on stack. You'll get rid of an extra my_free as a bonus

FooBarrior commented on extra/mariadbfrm/mariadb_frm.cc:
```diff
@@ -0,0 +1,531 @@
+  fflush(stdout);
+  
+
+  table= static_cast<TABLE *>(
```
same as for TABLE_SHARE

FooBarrior commented on extra/mariadbfrm/mariadb_frm.cc:
```diff
@@ -0,0 +1,531 @@
+     init_thread_environment() ||
+     mysql_init_variables() ||
+     init_character_sets() ||
+     init_plugin_system_complete())  // This is the key addition
```
only mysql_init_variables can really fail. I think it's reasonable to convert the rest of the functions to void and move outside if

FooBarrior commented on extra/mariadbfrm/mariadb_frm.cc:
```diff
@@ -0,0 +1,531 @@
+  .type = MYSQL_STORAGE_ENGINE_PLUGIN,
+  .info = &mock_hton,
+  .name = "MOCK_ENGINE",
+  .author = "Mock Author",
```
if mock_plugin survives our refactorings, you can put your name here:)

FooBarrior reviewed (changes requested):
Now it looks directed to the right way.

I noticed there are some junk files included into PR -- pls rollback them.
Add a test: create a table with a real server and invoke your frm parser with MTR's `--exec`

FooBarrior commented on extra/mariadbfrm/frm_mocks.cc:
```diff
@@ -0,0 +1,210 @@
+
+static void init_frm_mock_handlerton()
+{
+  frm_mock_hton_struct.create= frm_mock_create_handler;
+  frm_mock_hton_struct.db_type= DB_TYPE_UNKNOWN;
+
+  frm_mock_hton_struct.flags= HTON_NO_FLAGS;
```
Put this all in frm_mock_hton_struct designated initializer.

FooBarrior commented on extra/mariadbfrm/mariadb_frm.cc:
```diff
@@ -0,0 +1,498 @@
+extern mysql_cond_t COND_start_thread;
+extern struct system_variables global_system_variables;
+
+PSI_mutex_key key_LOCK_start_thread= 0;
+PSI_mutex_key key_LOCK_status= 0;
+PSI_mutex_key key_LOCK_global_system_variables= 0;
+PSI_mutex_key key_LOCK_user_conn= 0;
+PSI_mutex_key key_LOCK_thread_id= 0;
+PSI_cond_key key_COND_start_thread= 0;
+PSI_mutex_key key_LOCK_plugin= 0;
```
do we really need these? I think they are already initialized to zeroes in mysqld.cc.

hp77-creator replied:
`key_LOCK_thread_id` and `key_LOCK_plugin` are required, others I will remove.

FooBarrior reviewed (changes requested):
Overall looks good, I just commented a few creeps about the style, and I think many declarations might be redundand. (I guess that could be left from our past trials with bringing up the minimal linkage...)

You can test it all as follows:
```
create table t (x int); # will be executed by server
 --exec $MARIADB_FRM $MYSQL_TEST_DIR/std_data/frm/table_simple.frm
drop table t;

# etc, etc, etc...
```
to test different storage engine, add in the beginning:
```
--source include/have_innodb.inc
--source include/have_partition.inc
# and for example
--source include/have_blackhole.inc
```

What needs to be tested:

*  simple columns, at least int, text, varchar(value)
*  keys, `using btree`, `using rtree`
*  sinple vcol expressions: `a int, x int as (a *2)`
*  same for DEFAULT: `a int default (whatever)`
*  DEFAULT CURRENT_TIMESTAMP is a special case
*  field-level comments, table-level comments

Later on:
*  innodb engine
*  more exotic engines
*  engine options
*  Partitions!!

Later on:
* Expressions from plugins, like json, Item_geometry_func, inet6
* sequences: default(nextval(s))

Later on:
* Foreign keys in innodb

This is the full roadmap remaining, as I see it. I think we won't reach full support for everythink written in time, but at least:
* We shouldn't crash on those cases.
* We need to add an information on what's not supported in the commit message
* A test showing it's not crasing is still needed.

FooBarrior commented on extra/mariadbfrm/mariadb_frm.cc:
```diff
@@ -0,0 +1,498 @@
+
+static int mysql_init_variables()
+{
+  global_system_variables.character_set_client= &my_charset_utf8mb3_general_ci;
```
alternativelty, 
```
global_system_variables = { 
 // designited initializers
};
```
should work in c++20. Worth checking? 🤔

FooBarrior reviewed (changes requested):
Thank you for your work @hp77-creator!


I really like a steady collection of tests.

About the code - I see some functions handle unexistent error codes, etc, making me think that it's AI generated. If that's so, it should be reverted.

_AI is nothing bad in general, but the author should know what are they doing and validate the output._

For supporting the arbitrary expressions, I suggest you examine the following approach

1. Inherit an own Item, like Frm_parser_item
2. This item will do nothing but output an expression string it was supplied with through a constructor, in Frm_parser_item::print()
3. In unpack_vcol_info_from_frm:
#ifdef FRM_PARSER
return new Virtual_column_info(new Frm_parser_item(expr_str->c_ptr_safe(), expr_str->length())));
#endif
3.1. To make it work, youll have to add table.cc explicitly in the sources list! So that the macro would take effect.
3.2 Alternatively, make a global variable denoting an frm parser mode
3.3 Alternatively again, open_table_from_share->...->unpack_vcol_info_from_frm call chaing with an extra argument, denoting frm parser mode. This is the cleanest solution out of proposed.

Should work!

About the storage engines -- given that we already have a working mockup, an arbitrary engine should just work. The question is only in how to test it. 

For expample, EXAMPLE storage engine can be enabled with supplying `-DWITH_EXAMPLE_STORAGE_ENGINE=1` in the cmake line.

Then have_example_plugin.inc should work in tests after you'll build the target.

hp77-creator commented:
> About the code - I see some functions handle unexistent error codes, etc, making me think that it's AI generated. If that's so, it should be reverted.

> AI is nothing bad in general, but the author should know what are they doing and validate the output.

100% agree, I didn't validate everything it gave and just pushed it, sorry for that, I will go through the comments and fix whatever's hallucinated.

hp77-creator commented on extra/mariadbfrm/mariadb_frm.cc:
```diff
@@ -0,0 +1,498 @@
+  default_charset_info= &my_charset_utf8mb3_general_ci;
+
+  return 0;
+}
```
adding mysqld in include led to lot of linker errors so didn't change this much @FooBarrior

hp77-creator commented on extra/mariadbfrm/CMakeLists.txt:
```diff
@@ -0,0 +1,84 @@
+        )
+    ELSEIF(APPLE)
+    TARGET_LINK_OPTIONS(mariadb-frm PRIVATE
+            "-Wl,-multiply_defined,suppress"
```
i tried searching `ld` doesn't have replacement for `-multiply_defined`, trying with `-flat_namespace`

RazvanLiviuVarzaru commented:
Hi @hp77-creator !

`MTR` will complain about `mariadb-frm` when invoked after installing the packages as in: https://buildbot.mariadb.org/#/builders/889/builds/1596

```bash
mysql-test-run: *** ERROR: Could not find any of /usr/extra/mariadbfrm/mariadb-frm /usr/extra/mariadbfrm/mariadb-frm /usr/bin/mariadbfrm/mariadb-frm
```

I don't have a lot of server knowledge, but I'm sure @FooBarrior  can help you with this.

hp77-creator commented:
@FooBarrior shall I rebase all the commits into one for the final merge? All the CI are green now, one that is failing that is not because of my changes and it will pass if its re-triggered again.

FooBarrior commented:
@hp77-creator yes, please, make a single commit and rebase. Please mention in the commit what is implemented, and what is only tested.

Also mention our approach to reduce the size of the binary (so that there'd be an explanation for the linker flags)

FooBarrior commented:
@hp77-creator yes it's queued for adoption and merge

gkodinov reviewed (changes requested):
This seems to have fallen through and didn't get merged. I'm terribly sorry for that on behalf of the project. 
And I fully intend to get it moving again. 
I'd like to ask you one more thing before we proceed with the testing required: Can you please rebase on main? This is a new feature and it needs to go to that branch.

gkodinov reviewed (changes requested):
You did the rebase correctly. Unfortunately, when doing that UI alone won't do it (afaik). You also need to rebase your branch by yourself. 

Please ensure that the buildbot tests are passing. Right now there's a compilation error:

 ```
/home/buildbot/aarch64-debian-11/build/sql/table.cc: In function ‘bool parse_vcol_defs(THD*, MEM_ROOT*, TABLE*, bool*, vcol_init_mode, bool)’:
/home/buildbot/aarch64-debian-11/build/sql/table.cc:1273:96: error: too many arguments to function ‘Virtual_column_info* unpack_vcol_info_from_frm(THD*, TABLE*, String*, Virtual_column_info**, bool*)’
 1273 |                                     &((*field_ptr)->vcol_info), error_reported, frm_parser_mode);
```

gkodinov commented:
Do you intend to keep working on this? It's been 3 weeks since I've supplied review comments. Changing to draft due to inactivity. Please change it back when/if you are ready with the next batch.

hp77-creator commented on extra/mariadbfrm/frm_mocks.cc:
```diff
@@ -0,0 +1,263 @@
+    return pthread_cond_timedwait(cond, &mp->mutex, abstime);
+  }
+#endif // SAFE_MUTEX
+}
```
addressed

hp77-creator commented:
> Do you intend to keep working on this? It's been 3 weeks since I've supplied review comments. Changing to draft due to inactivity. Please change it back when/if you are ready with the next batch.

yes, I inted to get this merged. sorry for delay. I was occupied in some work stuff. will be more active here

FooBarrior commented:
@hp77-creator there is a (unintended) concatenation of various commit messages in your commit message, please check

hp77-creator commented:
> @hp77-creator there is a (unintended) concatenation of various commit messages in your commit message, please check

snap! I did multiple git rebase, probably that appended messages of every instant, i will `amend` it, thanks @FooBarrior

hp77-creator commented:
some tests are failing, converting to draft to resolve those. will make it RFR once done.

hp77-creator commented:
@gkodinov @FooBarrior please review.
@FooBarrior I have added support for vcols as well and removed conditional compilation for utility by using functional pointers.
Please let me know if this is alright or not.

hp77-creator committed:
Add FRM parser utility

- Parse FRM files
- Parse FRM files created in Engines like ARCHIVE, MEMORY, MyISAM, INNODB, CSV, ARIA
- Parse FRM files with default expressions like default(val)
- Parse FRM files that contain mathematical expressions like 1 * 2 * 3
- Parse FRM files containing sequences (exception are frm files of DDL expressions like CREATE SEQUENCE ...)
- Parse FRM files for some plugin functions like JSON (exception: INET and complex geometric functions are not yet supported)
- Provides a --debug option to stack trace in case of failures

Tests for following features are added:
- simple columns, at least int, text, varchar(value)
- keys, using btree
- keys using rtree
- simple vcol expressions: a int, x int as (a *2)
- same for DEFAULT: a int default (whatever)
- DEFAULT CURRENT_TIMESTAMP is a special case
- field-level comments, table-level comments
- innodb engine
- more exotic engines
- engine options
- partitions -- currently failing so added with error clause
- Expressions from plugins, like json, Item_geometry_func, inet6 -- currently failing so added with error clause
- Sequences: default(nextval(s))
- Foreign keys in innodb

- Applied aggressive linker flags to reduce executable size:
 - Used -Wl, --gc-sections to remove unused sections
 - Applied, -Wl,--strip-all to remove debugging symbols
 - Used -fno-common to remove duplicate symbol error

Reusing existing functions from table.cc to implement this FRM Parser util

gkodinov commented on extra/mariadbfrm/CMakeLists.txt:
```diff
@@ -0,0 +1,90 @@
+                -fno-common  
+        )
+        SET_TARGET_PROPERTIES(mariadb-frm PROPERTIES
+                LINK_FLAGS "-Wl,-w -Wl,-no_pie -Wl,-dead_strip -mmacosx-version-min=14.0"
```
Are you absolutely positive you need all of these options? E.g. -mmacosx-version-min has a cmake equivalent: CMAKE_OSX_DEPLOYMENT_TARGET. 
I'd strongly suggest to either remove all of these or reuse it from some other places.

gkodinov commented on extra/mariadbfrm/CMakeLists.txt:
```diff
@@ -0,0 +1,90 @@
+        )
+    ELSE()
+        TARGET_LINK_OPTIONS(mariadb-frm PRIVATE -flto
+                "LINKER:--allow-multiple-definition"
```
LINKER:--allow-multiple-definition seems like silencing the messenger to me. You most probably are specifying libraries twice or source files twice. 
I'd rather try to see what are the duplicates and attack that.

gkodinov commented on extra/mariadbfrm/CMakeLists.txt:
```diff
@@ -0,0 +1,90 @@
+        ${CMAKE_BINARY_DIR}/sql
+        ${CMAKE_BINARY_DIR}/include
+        ${LIBFMT_INCLUDE_DIR}
+        ${PCRE_INCLUDE_DIRS}
```
are you using PCRE? Why do you need to include this?

gkodinov commented on extra/mariadbfrm/CMakeLists.txt:
```diff
@@ -0,0 +1,90 @@
+    INCLUDE_DIRECTORIES(
+        ${CMAKE_SOURCE_DIR}/include
+        ${CMAKE_SOURCE_DIR}/sql
+        ${CMAKE_BINARY_DIR}/sql
```
why do you need to include /sql and /include twice, but in different order?

gkodinov commented on extra/mariadbfrm/CMakeLists.txt:
```diff
@@ -0,0 +1,90 @@
+        ${CMAKE_SOURCE_DIR}/sql
+        ${CMAKE_BINARY_DIR}/sql
+        ${CMAKE_BINARY_DIR}/include
+        ${LIBFMT_INCLUDE_DIR}
```
are you using libfmt in your utility? I sincerely doubt that.

gkodinov commented on extra/mariadbfrm/mariadb_frm.cc:
```diff
@@ -0,0 +1,658 @@
+  } while (0)
+
+extern mysql_mutex_t LOCK_thread_id;
+extern handlerton *get_frm_mock_handlerton();
```
this is very bad style! please put these in common header(s) so that they're actually checked against their definitions.

gkodinov commented on extra/mariadbfrm/CMakeLists.txt:
```diff
@@ -0,0 +1,90 @@
+IF(NOT WITHOUT_SERVER)
```
you will need a copyright header I'd guess.

gkodinov commented on extra/mariadbfrm/frm_mocks.cc:
```diff
@@ -0,0 +1,253 @@
+#include "handler.h"
```
this one will also need a copyright header.

gkodinov commented on sql/handler.cc:
```diff
@@ -264,6 +271,8 @@ handlerton *ha_default_tmp_handlerton(THD *thd)
 plugin_ref ha_resolve_by_name(THD *thd, const LEX_CSTRING *name,
                               bool tmp_table)
 {
+  if (ha_resolve_by_name_hook)
```
This feels like a very bad idea. As it will be done for the actual server in an actual open context every time you open a table.

gkodinov reviewed (approved):
Thank you for working on this. It's a very interesting idea. 
I'm somewhat terrified by the approach taken though: starting a gutted server with hooks that sway otherwise deeply layered code towards doing different things. But this is not for me to decide on. 

I'm guessing that, if approved, this needs a lot of cleanup. I've highlighted some of it below. 

But I'd strongly suggest considering getting an approval for the architecture first before we dive any deeper. 

Formal preliminary review criteria met, hence approving it.
